ingester2/src/server/grpc/query.rs

```diff
@@ -335,6 +347,72 @@ fn encode_response(
     })
 }
 
+#[derive(Debug)]
+enum StreamRecorderState {
+    Done,
+    Polling(Option<Box<Span>>),
+}
+
+struct StreamRecorder {
+    inner: FlightDataEncoder,
+    state: StreamRecorderState,
+    recorder: SpanRecorder,
+}
+
+impl StreamRecorder {
+    pub fn new(inner: FlightDataEncoder, span: Option<Span>, title: &'static str) -> Self {
+        let recorder = SpanRecorder::new(span);
+        Self {
+            inner,
+            state: StreamRecorderState::Polling(recorder.child_span(title).map(Box::new)),
```

Comment:
Perhaps we drop the `title` param (we won't ever change it at call time) and make the thing it's recording a little more explicit:

```suggestion
            state: StreamRecorderState::Polling(recorder.child_span("frame encoding").map(Box::new)),
```

That way the spans will be labelled with what they're doing - otherwise it'd just say "per partition" for every frame (which is per partition! confusing :fire:)